rre-core/src/main/java/io/sease/rre/core/domain/metrics/impl/NDCGAtK.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.sease.rre.core.domain.metrics.impl;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.sease.rre.core.domain.metrics.Metric;
+import io.sease.rre.core.domain.metrics.ValueFactory;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+import static io.sease.rre.Func.gainOrRatingNode;
+import static java.util.stream.Collectors.groupingBy;
+
+/**
+ * NDCG@k metric.
+ *
+ * @author agazzarini
+ * @since 1.0
+ */
+public class NDCGAtK extends Metric {
+    private final static BigDecimal TWO = new BigDecimal(2);
+    final int k;
+
+    /**
+     * Builds a new NDCGAtK metric.
+     *
+     * @param k the top k reference elements used for building the measure.
+     */
+    public NDCGAtK(@JsonProperty("k") final int k) {
+        super("NDCG@" + k);
+        this.k = k;
+    }
+
+    @Override
+    public int getRequiredResults() {
+        return k;
+    }
+
+    @Override
+    public ValueFactory createValueFactory(final String version) {
+        return new ValueFactory(this, version) {
+            private BigDecimal dcg = BigDecimal.ZERO;
+
+            @Override
+            public void collect(final Map<String, Object> hit, final int rank, final String version) {
+                if (rank > k) return;
+                judgment(id(hit))
+                        .ifPresent(judgment -> {
+                            final BigDecimal value = gainOrRatingNode(judgment).map(JsonNode::decimalValue).orElse(TWO);
```

Comment:
should value be "ratingValue" , should TWO be extracted as a constant called :

defaultRating, with a comment that says, this is the default rating used if the search result has no rating?

Reply:
I agree for the variable name, but once applied that change, I believe the comment is useless because the code clearly communicates the intent: 

```java                            
final BigDecimal ratingValue =     
             gainOrRatingNode(judgment)
                    .map(JsonNode::decimalValue)
                    .orElse(DEFAULT_RATING);
```